### ISSUE TITLE
pyrobuf should properly implement nested message definitions

### ISSUE THREAD

**Issue by daa** [appnexus/pyrobuf#55]
Currently pyrobuf fails when we reference one nested message definition from another.

Let's look at following proto definition:

```
message M {
    optional string s = 1;

    message N {
        required int32 value = 1;
    }

    message N2 {
        optional N nested_deeper = 1;
    }

    optional N2 nested = 2;
}
```

Here nested message N2 uses nested definition N defined at the same containing message M.

google protoc succeeds:

```
$ protoc nested.proto --python_out=.
$ echo $?
0
```

pyrobuf fails:

```
$ python -m pyrobuf --out-dir y x/nested.proto 
generating x/nested.proto
Compiling y/nested_proto.pyx because it changed.
[1/1] Cythonizing y/nested_proto.pyx

Error compiling Cython file:
------------------------------------------------------------
...
    cpdef void _Modified(self)

cdef class MN2:


    cdef N _nested_deeper
        ^
------------------------------------------------------------

y/nested_proto.pxd:43:9: 'N' is not a type identifier

Error compiling Cython file:
------------------------------------------------------------
...
        self._listener = <PyObject *>null_listener

    def __dealloc__(self):
        # Remove any references to self from child messages or repeated fields
        if self._nested_deeper is not None:
            self._nested_deeper._listener = <PyObject *>null_listener
                                           ^
------------------------------------------------------------

y/nested_proto.pyx:436:44: Cannot convert 'PyObject *' to Python object

Error compiling Cython file:
------------------------------------------------------------
...

    cpdef void reset(self):
        # reset values and populate defaults

        if self._nested_deeper is not None:
            self._nested_deeper._listener = <PyObject *>null_listener
                                           ^
------------------------------------------------------------

y/nested_proto.pyx:467:44: Cannot convert 'PyObject *' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
    cpdef void reset(self):
        # reset values and populate defaults

        if self._nested_deeper is not None:
            self._nested_deeper._listener = <PyObject *>null_listener
        self._nested_deeper = N.__new__(N)
                              ^
------------------------------------------------------------

y/nested_proto.pyx:468:31: undeclared name not builtin: N

Error compiling Cython file:
------------------------------------------------------------
...

        if self._nested_deeper is not None:
            self._nested_deeper._listener = <PyObject *>null_listener
        self._nested_deeper = N.__new__(N)
        self._nested_deeper.reset()
        self._nested_deeper._listener = <PyObject *>self
                                       ^
------------------------------------------------------------

y/nested_proto.pyx:470:40: Cannot convert 'PyObject *' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
        def __get__(self):
            return self._nested_deeper

        def __set__(self, value):
            if self._nested_deeper is not None:
                self._nested_deeper._listener = <PyObject *>null_listener
                                               ^
------------------------------------------------------------

y/nested_proto.pyx:480:48: Cannot convert 'PyObject *' to Python object

Error compiling Cython file:
------------------------------------------------------------
...

        def __set__(self, value):
            if self._nested_deeper is not None:
                self._nested_deeper._listener = <PyObject *>null_listener
            self._nested_deeper = value
            self._nested_deeper._listener = <PyObject *>self
                                           ^
------------------------------------------------------------

y/nested_proto.pyx:482:44: Cannot convert 'PyObject *' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
            self._Modified()

    cpdef void ClearField(self, field_name):
        """Clears the contents of a given field."""
        if field_name == 'nested_deeper':
            self._nested_deeper._listener = <PyObject *>null_listener
                                           ^
------------------------------------------------------------

y/nested_proto.pyx:517:44: Cannot convert 'PyObject *' to Python object

Error compiling Cython file:
------------------------------------------------------------
...
        """Clears the contents of a given field."""
        if field_name == 'nested_deeper':
            self._nested_deeper._listener = <PyObject *>null_listener
            self._nested_deeper = N.__new__(N)
            self._nested_deeper.reset()
            self._nested_deeper._listener = <PyObject *>self
                                           ^
------------------------------------------------------------

y/nested_proto.pyx:520:44: Cannot convert 'PyObject *' to Python object
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/daa/pyrobuf/pyrobuf/__main__.py", line 90, in <module>
    main()
  File "/home/daa/pyrobuf/pyrobuf/__main__.py", line 18, in main
    gen_message(args.source, out=args.out_dir, build=args.build_dir, install=args.install)
  File "/home/daa/pyrobuf/pyrobuf/__main__.py", line 68, in gen_message
    include_path=[os.path.join(HERE, 'src'), out]),
  File "/usr/lib64/python2.7/site-packages/Cython/Build/Dependencies.py", line 877, in cythonize
    cythonize_one(*args)
  File "/usr/lib64/python2.7/site-packages/Cython/Build/Dependencies.py", line 997, in cythonize_one
    raise CompileError(None, pyx_file)
Cython.Compiler.Errors.CompileError: y/nested_proto.pyx
```

because it uses class N which is not defined instead of generated MN, excerpt taken from nested.pxd:

```
cdef class MN2:


    cdef N _nested_deeper
```

similar in nested.pyx

**Comment by AndreMiras** [comment]
I was wondering, if we couldn't try to remove all the "is_nested" references as a "simple fix".
The idea would be done by not prefixing the "parent" message at all within the current message.
So removing all the "parent_name + current_name" concatenations in templates.
That way referencing to a type would always be as the type name, without having to rebuild its full hierarchy.
Would it make sense?
And by the way, with the current implementation, we only partly handle a hierarchy depth of 2 rather than being n.
One drawback I see with the implementing I suggest is it would lead to conflicting types in case of two message types with the same name, but one below another message. But such proto file would be misleading anyway.

**Comment by tburmeister** [comment]
That's not a bad idea, @AndreMiras, but I think having some form of namespacing/scope is important, just for the reason you point out with potentially conflicting types. I think we should try to find a way to generalize, perhaps by getting rid of the `is_nested` field, which is definitely a kludge,  and defining the name of the type in the internal representation as the name in the spec prefixed by the parent message name.

**Comment by AndreMiras** [comment]
I agree for the namespacing/scope. Yes we could try to keep a track of the type hierarchy while parsing the proto file in parse_proto.py. But I'm not too sure how we would implement that.
Do you have an idea?

**Comment by tburmeister** [comment]
I have the seed of an idea, but have not fleshed it out completely. For starters,  I think we should pass the "parent" message to `_parse_message` and start off that function with the following:
`current.messages.update(parent.messages)
current.enums.update(parent.enums)`
That at least gives us recursive access to all messages "in scope". Still not totally sure on the best way to handle naming.

**Comment by AndreMiras** [comment]
OK thanks for the pointers, it makes sense. I'm not sure I'll find time and courage to work on it, but we never know :)

**Comment by AndreMiras** [comment]
I would like to take a more serious look, but there're few other things I'm seeing now.
Would that be reasonable if I started pull requesting by small increments?
For example I would like to:
1) Drop all the `is_nested` logic and implement the flat hierarchy first, making sure all tests are still passing. This would create a cleaner playground for next tasks.
2) Then I would like to correct a bunch of other minor things in a different pull request.
3) And then take a look at the proper nested implementation in another pull request.
Does it seem reasonable?
I'm ready to start straight away with (1), otherwise I'm open to suggestions. The only thing is, I don't feel like doing (1), (2) and (3) at once since it's a big task and I'm not sure I would have time and energy to finish.

**Comment by tburmeister** [comment]
I'm still hesitant to drop namespacing, even if temporarily. Interestingly, just making the following change to `_parse_message` gets the original example to compile:

```
def _parse_message(self, s, current, parent, tokens, enums, imported_enums):
        # recursive access to messages and enums defined in higher scopes
        if parent is not None:
            current.messages.update(parent.messages)
            current.enums.update(parent.enums)
```

The only issue is that it winds up generating a `MN` class and an identical `MN2N` class, which is obviously not ideal, but seems like a start.

**Comment by AndreMiras** [comment]
I have a working code for messages, I now need to update for enums, plus clean the code before submitting the pull request.
Rather than using the `parent` parameter for `_parse_message()`, I'm using a `messages` parameter that contains all known messages. Also the `ParserMessage` now contains a `full_name` attribute that "describes" the message full hierarchy name. I'm currently using a "ParentChild" naming style to keep the API backward compatible with older versions. But it could really be something like "Parent.Child". I think, this is what protoc is doing, but I'm not sure yet.
Last thing, is the `proto_pxd.tmpl` now uses recursive macro call, so it should work for "N" hierarchy levels, but `proto_pyx.tmpl` is not yet recursive. Let's see later.

**Comment by tburmeister** [comment]
Awesome! Heads up that Cython does not allow nested class definitions, so unfortunately we can't do `Parent.Child`.

**Comment by AndreMiras** [comment]
A first pull request is ready to to be reviewed here #66, all tests are passing so hopefully I didn't introduced any regression.

**Comment by tburmeister** [comment]
Believe this was fixed by #66
